format/postgres/common/pg_heap/postgres/pg_heap.go

```diff
@@ -0,0 +1,415 @@
+//
+/* total size (bytes):   24 */
+
+// type = struct PageXLogRecPtr {
+/*    0      |     4 */ // uint32 xlogid;
+/*    4      |     4 */ // uint32 xrecoff;
+
+/* total size (bytes):    8 */
+
+// typedef uint16 LocationIndex;
+// #define SizeOfPageHeaderData (offsetof(PageHeaderData, pd_linp))
+
+// type = struct HeapTupleHeaderData {
+/*    0      |    12 */ // union {
+/*                12 */ //     HeapTupleFields t_heap;
+/*                12 */ //     DatumTupleFields t_datum;
+//
+/* total size (bytes):   12  */
+/* } t_choice;             //
+/*   12      |     6 */ // ItemPointerData t_ctid;
+/*   18      |     2 */ // uint16 t_infomask2;
+/*   20      |     2 */ // uint16 t_infomask;
+/*   22      |     1 */ // uint8 t_hoff;
+/*   23      |     0 */ // bits8 t_bits[];
+/* XXX  1-byte padding  */
+//
+/* total size (bytes):   24 */
+const SizeOfHeapTupleHeaderData = 24
+
+// type = struct HeapTupleFields {
+/*    0      |     4 */ // TransactionId t_xmin;
+/*    4      |     4 */ // TransactionId t_xmax;
+/*    8      |     4 */ // union {
+/*                 4 */ //    CommandId t_cid;
+/*                 4 */ //    TransactionId t_xvac;
+//                         } t_field3;
+/*                      total size (bytes):    4 */
+//
+/* total size (bytes):   12 */
+
+// type = struct DatumTupleFields {
+/*    0      |     4 */ // int32 datum_len_;
+/*    4      |     4 */ // int32 datum_typmod;
+/*    8      |     4 */ // Oid datum_typeid;
+//
+/* total size (bytes):   12 */
+
+// type = struct ItemPointerData {
+/*    0      |     4 */ // BlockIdData ip_blkid;
+/*    4      |     2 */ // OffsetNumber ip_posid;
+//
+/* total size (bytes):    6 */
+
+type Heap struct {
+	Args format.PostgresHeapIn
+
+	// current Page
+	Page *HeapPage
+	// Page special data
+	Special *PageSpecial
+
+	// current tuple
+	Tuple *TupleD
+
+	DecodePageHeaderData func(page *HeapPage, d *decode.D)
+	DecodePageSpecial    func(heap *Heap, d *decode.D)
+}
+
+type PageSpecial struct {
+	// pgproee
+	PdXidBase   uint64 // 8 TransactionId pd_xid_base;
+	PdMultiBase uint64 // 8 TransactionId pd_multi_base;
+	PdPruneXid  uint64 // 4 ShortTransactionId pd_prune_xid;
+	PdMagic     uint64 // 4 uint32 pd_magic;
+}
+
+type TupleD struct {
+	IsMulti bool
+}
+
+func Decode(heap *Heap, d *decode.D) any {
+	decodeHeapPages(heap, d)
+	return nil
+}
+
+func decodeHeapPages(heap *Heap, d *decode.D) {
+	blockNumber := uint32(heap.Args.PageNumber + heap.Args.SegmentNumber*common.RelSegSize)
+	count := int64(0)
+	for {
+		if end, _ := d.TryEnd(); end {
+			return
+		}
+
+		d.FieldStruct("page", func(d *decode.D) {
+			decodeHeapPage(heap, d, blockNumber)
+		})
+		blockNumber++
+		count++
+
+		// end of Page
+		endLen := uint64(d.Pos() / 8)
+		pageEnd := int64(common.TypeAlign(common.PageSize, endLen))
+		pageEnd0 := count * common.PageSize
+		if pageEnd0 != pageEnd {
+			fmt.Printf("invalid page %d end expected %d, actual %d, endLen  %d\n", count-1, pageEnd0, pageEnd, endLen)
```

Comment:
`d.Fatalf` or `d.Errorf`?

Reply:
I don't want to interrupt execution here, just print warning.

Reply:
I see. Decoder code is not suppose to do any io like this and if they would they should probably do it thru some abstraction instead printing directly to stdout. So i wonder if we can come up with some other solution.

I see some possible ways:
- Make it possible to add warnings the the code tree somehow. I kind of can have errors already.
- Use description on some field to indicate something is wrong
- Populate with fields and provide some query in the documentation to find this condition

BTW if you use `d.Errorf` fq will by default fail decoding at the point but you can tell it to ignore the error by using `-o force=true`. `d.Fatalf` is if the error should not be possible to ignore.

Reply:
I changed fmt.Printf -> d.Errorf.